python/lsst/pipe/base/pipelineIR.py

```diff
@@ -430,6 +551,84 @@ def _read_tasks(self, loaded_yaml):
+        have issues, but it is the only feasible way when users can specify
+        contracts with generic strings.
+        """
+        new_contracts = []
+        for contract in self.contracts:
+            # match a label that is not preceded by an ASCII identifier, or
+            # is the start of a line and is followed by a dot
+            if re.match(f".*([^A-Za-z0-9_]|^){label}[.]", contract.contract):
+                continue
+            new_contracts.append(contract)
+        self.contracts = new_contracts
+
+    def subset_from_labels(self, labelSpecifier: Set[str]) -> PipelineIR:
+        """Subset a pipelineIR to contain only labels specified in labelSpecifier
+
+        Parameters
+        ----------
+        labelSpecifier : `set` of `str`
+            Set containing labels that describes how to subset a pipeline.
+
+        Returns
+        -------
+        pipeline : `PipelineIR`
+            A new pipelineIR object that is a subset of the old pipelineIR
+
+        Raises
+        ------
+        ValueError
+            Raised if there is an issue with specified labels
+
+        Notes
+        -----
+        This method attempts to prune any contracts that contain labels which
+        are not in the declared subset of labels. This pruning is done using a
+        string based matching due to the nature of contracts and may prune more
+        than it should.
+        """
+
+        pipeline = copy.deepcopy(self)
+
+        # update the label specifier to expand any named subsets
+        toRemove = set()
+        toAdd = set()
+        for label in labelSpecifier:
+            if label in pipeline.labeled_subsets:
+                toRemove.add(label)
+                toAdd.update(pipeline.labeled_subsets[label].subset)
+        labelSpecifier.difference_update(toRemove)
+        labelSpecifier.update(toAdd)
+        # verify all the labels are in the pipeline
+        if not labelSpecifier.issubset(pipeline.tasks.keys()
+                                       | pipeline.labeled_subsets):
+            difference = labelSpecifier.difference(pipeline.tasks.keys())
+            raise ValueError("Not all supplied labels (specified or named subsets) are in the pipeline "
+                             f"definition, extra labels: {difference}")
+        # copy needed so as to not modify while iterating
+        pipeline_labels = set(pipeline.tasks.keys())
+        # Remove the labels from the pipelineIR, and any contracts that contain
+        # those labels (see docstring on _remove_contracts for why this may
+        # cause issues)
+        for label in pipeline_labels:
+            if label not in labelSpecifier:
+                pipeline.tasks.pop(label)
+                pipeline._remove_contracts(label)
+
+        # create a copy of the object to iterate over
+        labeled_subsets = copy.copy(pipeline.labeled_subsets)
+        # remove any labeled subsets that no longer have a complete set
```

Comment:
This sounds like the right behavior, but it's sufficiently non-obvious that it should probably be mentioned in the docstring as well.